### PR TITLE
[WIP]feat: add network in use check to `compose down`

### PR DESCRIPTION
**Pull request by Retrospection**
To solve issue #1425

### CODE REVIEW

**Comment by yuchanns** [comment]
Please using a valid DCO Sign Off.

File: pkg/netutil/netutil.go [hunk 1]
```diff
@@ -49,37 +49,7 @@ type CNIEnv struct {
...
-			} else {
-				used[k] = v
-			}
-		}
-	}
-	return used, nil
-}
-
-func namespaceUsedNetworks(ctx context.Context, containers []containerd.Container) (map[string][]string, error) {
+func NamespaceUsedNetworks(ctx context.Context, containers []containerd.Container) (map[string][]string, error) {
```
Comment (Zheaoli) [inline pkg/netutil/netutil.go:52]: I think you just make `NamespaceUsedNetworks` public here, I think it's necessary to reorder the function

File: pkg/composer/down.go [hunk 2]
```diff
@@ -96,7 +96,18 @@ func (c *Composer) downNetwork(ctx context.Context, shortName string) error {
...
+	}
+
+	if netExists {
+		netUsed, err := c.NetworkUsed(ctx, fullName)
+		if err != nil {
+			return err
+		}
+
+		if netUsed {
+			return fmt.Errorf("failed to remove network %q: Error response from daemon: error while removing network: network %q has active endpoints", fullName, fullName)
```
Comment (Zheaoli) [inline pkg/composer/down.go:108]: The error message here could be the same with https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/network_rm.go#L76.

Just make the message as simple as we can

File: cmd/nerdctl/compose.go [hunk 3]
```diff
@@ -156,6 +157,38 @@ func getComposer(cmd *cobra.Command, client *containerd.Client) (*composer.Compo
...
+		}
+		used := make([]string, 0)
+		for _, ns := range nsList {
+			nsCtx := namespaces.WithNamespace(ctx, ns)
+			containers, err := client.Containers(nsCtx)
+			if err != nil {
+				return false, err
+			}
+			nsUsedN, err := netutil.NamespaceUsedNetworks(nsCtx, containers)
+			if err != nil {
```
Comment (Zheaoli) [inline cmd/nerdctl/compose.go:174]: I think we can make the code simple here,

you can make the code like this

```go
	usedNetworkInfo, err := netutil.UsedNetworks(ctx, client)
	if err != nil {
		return false, err
	}
        if _,ok:=usedNetworkInfo[netName];ok{
                return true, nil
        }
```

**Review by Zheaoli — CHANGES_REQUESTED** [review]
And we need extra e2e test, and plz sign-off your commit,

File: pkg/composer/down.go (hunk 2 above)
Comment (Retrospection) [reply pkg/composer/down.go:108]: make sense

> The error message here could be the same with https://github.com/containerd/nerdctl/blob/main/cmd/nerdctl/network_rm.go#L76.
> 
> Just make the message as simple as we can

File: pkg/composer/composer.go [hunk 4]
```diff
@@ -41,6 +41,7 @@ type Options struct {
 	EnvFile          string
 	NerdctlCmd       string
 	NerdctlArgs      []string
+	NetworkUsed      func(context.Context, string) (bool, error)
```
Comment (djdongjin) [inline pkg/composer/composer.go:44]: nit: suggest rename to `NetworkInUse` (similarly `netInUse` below)

Comment (Retrospection) [reply pkg/composer/composer.go:44]: make sense

**Pull request closed without merging**